core/base/batch_utilities.hpp

```diff
@@ -154,6 +162,241 @@ std::vector<gko::matrix_data<ValueType, IndexType>> write(
+    }
+
+    auto b_size = batch_dim<2>(num_batch_items, common_size);
+    size_type batch = 0;
+    std::vector<mat_data> input_mat_data(num_batch_items, common_size);
+    for (const auto& b : vals) {
+        size_type ridx = 0;
+        for (const auto& row : b) {
+            size_type cidx = 0;
+            for (const auto& elem : row) {
+                if (elem != zero<value_type>()) {
+                    input_mat_data[batch].nonzeros.emplace_back(ridx, cidx,
+                                                                elem);
+                }
+                ++cidx;
+            }
+            ++ridx;
+        }
+        ++batch;
+    }
+    return read<value_type, index_type, Matrix>(
+        exec, input_mat_data, std::forward<TArgs>(create_args)...);
+}
+
+
+/**
+ * Creates and initializes a batch single column-vector by making copies of the
+ * single input column vector.
+ *
+ * This function first creates a temporary batch multi-vector, fills it with
+ * passed in values, and then converts the vector to the requested type.
+ *
+ * @tparam Matrix  matrix type to initialize
+ *                 (MultiVector has to implement the ConvertibleTo<Matrix>
+ *                  interface)
+ * @tparam TArgs  argument types for Matrix::create method
+ *                (not including the implied Executor as the first argument)
+ *
+ * @param num_vectors  The number of times the input vector is to be duplicated
+ * @param vals  values used to initialize each vector in the temp. batch
+ * @param exec  Executor associated to the vector
+ * @param create_args  additional arguments passed to Matrix::create, not
+ *                     including the Executor, which is passed as the first
+ *                     argument
+ *
+ * @ingroup MultiVector
+ * @ingroup mat_formats
+ */
+template <typename Matrix, typename... TArgs>
+std::unique_ptr<Matrix> initialize(
+    const size_type num_vectors,
+    std::initializer_list<typename Matrix::value_type> vals,
+    std::shared_ptr<const Executor> exec, TArgs&&... create_args)
+{
+    using value_type = typename Matrix::value_type;
+    using index_type = typename Matrix::index_type;
+    using mat_data = gko::matrix_data<value_type, index_type>;
+    size_type num_batch_items = num_vectors;
+    GKO_THROW_IF_INVALID(num_batch_items > 0 && vals.size() > 0,
+                         "Input data is empty");
+    auto num_rows = begin(vals) ? vals.size() : 0;
+    auto common_size = dim<2>(num_rows, 1);
+    auto b_size = batch_dim<2>(num_batch_items, common_size);
+    std::vector<mat_data> input_mat_data(num_batch_items, common_size);
+    for (size_type batch = 0; batch < num_vectors; batch++) {
+        input_mat_data[batch].nonzeros.reserve(num_rows);
+        size_type idx = 0;
+        for (const auto& elem : vals) {
+            if (elem != zero<value_type>()) {
+                input_mat_data[batch].nonzeros.emplace_back(idx, 0, elem);
+            }
+            ++idx;
+        }
```

Comment:
create the mat_data first, and then use `std::vector<mat_data> input_mat_data(num_batch_items, mat_data)` to duplicate?